### ISSUE TITLE
Use github-actions-random-matrix to increase test coverage and reduce PR build durations

### ISSUE THREAD

**Issue by vlsi** [kotest/kotest#3389]
## Testing Problem

It would be nice to have fast feedback from PR builds (== reduce the number of executed CI jobs).
Currently, CI builds a lot of jobs which rarely fail, yet they spend contributors' time, and they are not really environmental-friendly as they burn CPU with little gain.

It would be great to increase the tested configuration space at the same time.
It would be nice to ensure kotest is buildable and workable with various OS, JDKs, timezone, etc, settings.

## Suggested Solution

For example, the automatic matrix captures locale-dependent issues like https://github.com/spockframework/spock/issues/1414
See how the matrix looks in:
* https://github.com/pgjdbc/pgjdbc/actions/runs/4101968886
* https://github.com/cbeust/testng/actions/runs/4077882114
    
The idea is that you declare test axis like `os`, `java`, `java_distribution`, `timezone`, `locale`, etc, and `matrix.js` produces several jobs for you.

See https://github.com/vlsi/github-actions-random-matrix
It enables a clean and concise way to configure and maintain the test matrix.

The PR could be similar to https://github.com/pgjdbc/pgjdbc/pull/2534

WDYT?

**Comment by sksamuel** [comment]
If someone introduces a PR that breaks TVOS build, but nothing else, and that platform happens to be skipped in the PR, how do we know? We wouldn't until some point in the future when it was included, and who then fixes it? Probably me or @Kantis and that just adds to our workload.

So I don't agree with not running every platform. The other axis make more sense for randomizing. And as for env-concerns - I'd rather run the builds once up front than multiple times in the future trying to find why it was failing.

**Comment by vlsi** [comment]
> We wouldn't until some point in the future when it was included, and who then fixes it? Probably me or @Kantis and that just adds to our workload.

You can always tag the original contributor.

You can run a broader set of tests on the main branch, so you'll notice the failure immediately after you merge the changes, so your "wouldn't until some point in the future" argument is not justified.

**Comment by Kantis** [comment]
I think it's good that we run tests for all platforms. They run in parallel anyway, so wouldn't be much gained from excluding some?

If you want faster feedback, why not just run the tests locally for the platform you're interested in?

**Comment by vlsi** [comment]
>They run in parallel anyway, so wouldn't be much gained from excluding some?

The server resources are finite.
Here's a case when PR waited for 1 hour in the queue: https://github.com/kotest/kotest/pull/3365#issuecomment-1412189655

It does distract the contributors.

>They run in parallel anyway, so wouldn't be much gained from excluding some?

Excluding some jobs would allow more room for other jobs.

**Comment by sksamuel** [comment]
> > They run in parallel anyway, so wouldn't be much gained from excluding some?
> 
> The server resources are finite. Here's a case when PR waited for 1 hour in the queue: [#3365 (comment)](https://github.com/kotest/kotest/pull/3365#issuecomment-1412189655)
> 
> It does distract the contributors.

You get the build green locally then push it up for the full test suite.

I'm sorry I don't want to remove all the builds. It will come back to bite me and I'll end up debugging an environment I know nothing about (mac or windows).

**Comment by vlsi** [comment]
Ok, I have no more interest in discussing this. I unsubscribe. Feel free to do whatever you want.
